ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,152 @@
+- If the number of successful BuildRuns is larger than `successfulLimit`, then the oldest successful BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+- If the number of failed BuildRuns is larger than `failedLimit`, then the oldest failed BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+
+We can consider to run above logic in a go routine rather than as part of the BuildRun reconciliation. Failures would be ignored which is okay.
+
+When the `ttlAfterFailed` field is set for a BuildRun that failed, or when the `ttlAfterSuccessful` field is set for a BuildRun that succeeded, then we will return a reconciliation result where `Requeue` is set to `true` and `RequeueAfter` is set to `.status.completionTime + TTL - now`.
+
+When a controller becomes the leader, a `CreateEvent` happens for all BuildRuns. [Today, we only reconcile](https://github.com/shipwright-io/build/blob/v0.8.0/pkg/reconciler/buildrun/controller.go#L57-L63) if the BuildRun has not yet a TaskRun reference and has not yet completed. We will also need to trigger a reconcile in case it has completed and there is a TTL for the BuildRun.
+
+Our BuildRun reconciliation logic will need to be extended. [After we fetched the BuildRun](https://github.com/shipwright-io/build/blob/v0.8.0/pkg/reconciler/buildrun/buildrun.go#L82), we will need to add logic that checks if the BuildRun is already completed and check if we need to delete it or requeue for deletion.
+
+TBD When a Build gets modified and any value inside the `retention` section gets initially set or lowered, then ... [see planned discussion in community meeting](https://github.com/shipwright-io/community/issues/62#issuecomment-1041378397).
+
+### Staging
+
+The implementation can be done in stages. In particular, the TTL-based cleanup and the limit-based cleanup are very much independent.
+
+### Documentation updates
+
+We will need to update the documentation about builds to introduce this new feature.
+
+### User Stories
+
+#### As a Build user, I want to define how many finished BuildRuns I want to keep for my Build so that the amount of finished BuildRuns does not grow endlessly
+
+#### As a Build user, I want to define how long a finished BuildRun should stay so that I do not have obsolete and very old BuildRuns in the system
+
+### Test Plan
+
+It should include:
+
+- Integration tests
+- Unit tests
+- e2e tests
+
+### Release Criteria
+
+Should be available before release v1.0.0
+
+### Risks and Mitigations
+
+We will have more logic in our reconciliations. Especially the required `list` calls can become a problem if `successfulLimit`, or `failedLimit` are set to high values and we consequently retrieve a large amount of BuildRuns from the API server. A mitigation would be that we define maximum values for those values (maybe 100).
```

Comment:
+1, this contradicts the one-shot previous efforts, due to extra reconcile loops.